src/ImageSharp/Processing/Processors/Transforms/Linear/AffineTransformProcessor{TPixel}.cs

```diff
@@ -129,100 +110,119 @@ public void ApplyTransform<TResampler>(in TResampler sampler)
+
                 Buffer2D<TPixel> sourceBuffer = this.source.PixelBuffer;
 
-                PixelOperations<TPixel>.Instance.ToVector4(
-                    this.configuration,
-                    this.destination.GetPixelRowSpan(y),
-                    span);
+                for (int y = rows.Min; y < rows.Max; y++)
+                {
+                    Span<TPixel> rowSpan = this.destination.GetPixelRowSpan(y);
+                    PixelOperations<TPixel>.Instance.ToVector4(
+                        this.configuration,
+                        rowSpan,
+                        span,
+                        PixelConversionModifiers.Scale);
 
-                ref float yKernelSpanRef = ref MemoryMarshal.GetReference(this.yKernelBuffer.GetRowSpan(y));
-                ref float xKernelSpanRef = ref MemoryMarshal.GetReference(this.xKernelBuffer.GetRowSpan(y));
+                    for (int x = 0; x < span.Length; x++)
+                    {
+                        var point = Vector2.Transform(new Vector2(x, y), matrix);
+                        float pY = point.Y;
+                        float pX = point.X;
+
+                        int top = LinearTransformUtility.GetRangeStart(yRadius, pY, maxY);
+                        int bottom = LinearTransformUtility.GetRangeEnd(yRadius, pY, maxY);
+                        int left = LinearTransformUtility.GetRangeStart(xRadius, pX, maxX);
+                        int right = LinearTransformUtility.GetRangeEnd(xRadius, pX, maxX);
+
+                        if (bottom == top || right == left)
+                        {
+                            continue;
+                        }
+
+                        Vector4 sum = Vector4.Zero;
+                        for (int yK = top; yK <= bottom; yK++)
+                        {
+                            float yWeight = sampler.GetValue(yK - pY);
+
+                            for (int xK = left; xK <= right; xK++)
+                            {
+                                float xWeight = sampler.GetValue(xK - pX);
+
+                                Vector4 current = sourceBuffer.GetElementUnsafe(xK, yK).ToScaledVector4();
+                                Numerics.Premultiply(ref current);
+                                sum += current * xWeight * yWeight;
+                            }
+                        }
+
+                        span[x] = sum;
+                    }
 
-                for (int x = 0; x < this.maxX; x++)
-                {
-                    // Use the single precision position to calculate correct bounding pixels
-                    // otherwise we get rogue pixels outside of the bounds.
-                    var point = Vector2.Transform(new Vector2(x, y), this.matrix);
-                    LinearTransformUtils.Convolve(
-                        in this.sampler,
-                        point,
-                        sourceBuffer,
+                    Numerics.UnPremultiply(span);
```

Comment:
Have we used the bulk `UnPremultiply` in the old code?

Reply:
We didn't. Have just ran a test commit using the non-bulk variant and the major issues disappeared leaving only 2 minor (_max 0.03% diff_) remaining. 

Below is the output of the worst issue. The last 3 pixels of each row appear to be affected which suggests to me that something is going wrong here. My guess is that it has something to do with unpremultiplying Vector4.Zero 

https://github.com/SixLabors/ImageSharp/blob/bb727699686cba9cf55b67c5b114bb45311e6f3f/src/ImageSharp/Common/Helpers/Numerics.cs#L556-L576

![image](https://user-images.githubusercontent.com/385879/113943236-26974c80-97fa-11eb-81dc-03e44b8411e2.png)

The question is, what do we do? The issue is clearly fixed for .NET 5 but .NET Core 3.1 is LTS.

Reply:
I think we should still keep the bulk code for the rest of the platforms, because it likely takes an important part in the perf improvements we see in your benchmarks. I would add a check for the OS platform at this line with an explanatory comment:

https://github.com/SixLabors/ImageSharp/blob/bb727699686cba9cf55b67c5b114bb45311e6f3f/src/ImageSharp/Common/Helpers/Numerics.cs#L543

```C#
if (Avx2.IsSupported && RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && vectors.Length >= 2)
```

In no case I would complicate this with additional logic to detect if we are running on .NET 5 (Mac is not *that* important).